From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manish Bhardwaj <mbhardwaj@nvidia.com>
Date: Tue, 19 Aug 2025 14:59:38 +0000
Subject: [PATCH 193/241] ivc: cert & misra fixes in ivc.c drv

Fixing below cert & misra errors:-
- CID 27344024: cert_int08_c_violation: Unsigned integer operation
  offset + size may wrap.
- CID 27344023: cert_int30_c_violation: Unsigned integer operation
  offset + size may wrap

Bug 5433066

Change-Id: Ief4d506cb1564064806d373e4e3beeb210fb6e21
Signed-off-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3449490
Reviewed-by: Sandeep Trasi <strasi@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Suresh Venkatachalam <skathirampat@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
(cherry picked from commit d4453e5a177ea56ca13f21ec899d95374e91d8f8)

diff --git a/drivers/firmware/tegra/ivc.c b/drivers/firmware/tegra/ivc.c
index e4745c08ff8f..7d7956e0a1d1 100644
--- a/drivers/firmware/tegra/ivc.c
+++ b/drivers/firmware/tegra/ivc.c
@@ -674,6 +674,7 @@ int tegra_ivc_read_peek(struct tegra_ivc *ivc, void __user *usr_buf,
 {
 	struct iosys_map map;
 	int err;
+	size_t result;
 
 	WARN_ON_ONCE(buf && usr_buf);
 
@@ -683,10 +684,14 @@ int tegra_ivc_read_peek(struct tegra_ivc *ivc, void __user *usr_buf,
 		return err;
 
 	/* Ensure offset + size does not exceed buffer bounds */
-	if (offset + size > ivc->frame_size) {
+	if (check_add_overflow(offset, size, &result)) {
+		pr_err("%s: operation got overflown.\n", __func__);
 		return -EINVAL;
 	}
 
+	if (result > ivc->frame_size)
+		return -EINVAL;
+
 	/* update the buffer with read data*/
 	if (buf) {
 		iosys_map_memcpy_from(buf, &map, offset, size);
-- 
2.43.0

